### PR TITLE
Don't allow CSE & Constant Pooling to introduce new aliasing of graph outputs

### PR DESCRIPTION
**Pull request by eellison**
Since we don't know whether the graph outputs will be mutated once they are returned, CSE and constant pooling cannot introduce new aliasing relationships. This can occur in autodiff subgraph slicing for example, or in a python op that calls into a scripted op. 

The first commit is a fix for alias analysis where we always considered Container Types to have a fresh alias. We need a value and a tuple constructed of that value to alias each other to check for aliasing in graph outputs. 

Second commit is the change to constant pooling & cse. 

This is the replacement of https://github.com/pytorch/pytorch/pull/18491

### CODE REVIEW

**Comment by eellison** [comment]
after talking with @zdevito in person, we're going to wait until michael's off PTO and land just the fix for subgraph slicing for now

File: torch/csrc/jit/passes/common_subexpression_elimination.cpp
Comment (suo) [inline torch/csrc/jit/passes/common_subexpression_elimination.cpp]: don't define this twice?

**Review by suo — CHANGES_REQUESTED** [review]
This PR is explicitly two changes glommed together. In the future please stack two separate PRs so we can discuss the changes individually.

File: torch/csrc/jit/passes/common_subexpression_elimination.cpp
Comment (eellison) [reply torch/csrc/jit/passes/common_subexpression_elimination.cpp]: It's a relatively small function - where do you want me to put this?

**Comment by eellison** [comment]
Closing until https://github.com/pytorch/pytorch/pull/18710 passes

**Pull request closed without merging**
